### PR TITLE
feat(middleware-correlation-ids): support sns sqs without raw deliveries

### LINKED ISSUE #23: Handle correlation IDs which are nested via SNS => SQS => Lambda integration
# This is a Feature Proposal

## Description

* When integrating SNS (published with correlation IDs) with SQS, the correlationIds middleware only looks for messageAttributes on the SQS message
* If no correlationIds are present on the SQS message, we could try to extract them from the record.body and use the SNS message correlation IDs

Pseudo-code:
```js
if (!sqsCorrelationIds) {
  const body = JSON.parse(record.body);
  if (body.MessageAttributes) {
    record.correlationIds = extractSNSCorrelationIds(body.MessageAttributes)
  }
}

### PR DESCRIPTION
**Pull request by gsingh1**
## What did you implement:

Closes #23 

Handle correlation IDs which are nested via SNS => SQS => Lambda integration in the middleware correlation ids package.
When raw message delivery is disabled in the SNS subscription, messages are delivered in JSON format along with SNS metadata along with the body of the message itself.

## How did you implement it:

If message attributes are not found in the SQS message payload, try to look for an SNS message payload and check if message attributes are set there.

## How can we verify it:

Create an SNS subscription with raw message delivery disabled with an SQS integration type which is has a lambda trigger. The original SNS message should have an correlation ID set in the message attributes which can then be picked up by the middleware now.

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / projects / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

File: packages/lambda-powertools-middleware-correlation-ids/__tests__/index.js [hunk 1]
```diff
@@ -455,7 +468,9 @@ describe('Correlation IDs middleware', () => {
 
   describe('SFN', () => standardTests(genSfnEvent))
 
-  describe('SQS', () => sqsTests())
+  describe('SQS', () => sqsTests(false))
+
+  describe('SQS Without Raw Message', () => sqsTests(true))
```
[the remark this answers is not in the document]
Comment (gsingh1) [reply packages/lambda-powertools-middleware-correlation-ids/__tests__/index.js]: tests now added to ensure message attributes are not overridden

File: packages/lambda-powertools-middleware-correlation-ids/index.js [hunk 2]
```diff
@@ -70,6 +70,26 @@ function captureSqs (event, context, sampleDebugLogRate) {
...
+      } catch (e) {
+      }
+
+      if (body.hasOwnProperty('MessageAttributes') &&
+        body.hasOwnProperty('TopicArn') &&
+        body.TopicArn.startsWith('arn:aws:sns')
+      ) {
+        for (const bodyMsgAttribute in body.MessageAttributes) {
+          const stringValue = body.MessageAttributes[bodyMsgAttribute].Value
+          msgAttributes[bodyMsgAttribute] = { stringValue }
```
[the remark this answers is not in the document]
Comment (gsingh1) [reply packages/lambda-powertools-middleware-correlation-ids/index.js]: Good spot, now fixed 👌

File: packages/lambda-powertools-middleware-correlation-ids/__tests__/index.js [hunk 3]
```diff
@@ -457,5 +498,7 @@ describe('Correlation IDs middleware', () => {
 
   describe('SQS', () => sqsTests())
 
+  describe('SQS Without Raw Message', () => sqsWithoutRawDeliveryTests())
```
Comment (theburningmonk) [inline packages/lambda-powertools-middleware-correlation-ids/__tests__/index.js]: I'm probably just nick-picking here, but I'm not fond of the `SQS without raw message` name as the distinction of raw message or not is specific to SNS => SQS. Can we try a more meaningful name? e.g. "wrapped SNS message", "nested SNS message", "SNS message"?

Comment (gsingh1) [reply packages/lambda-powertools-middleware-correlation-ids/__tests__/index.js]: test names now updated to `SQS wrapped SNS message`

**Pull request merged**